### DISCUSSION TITLE
xsns_86_tfminiplus.ino

### DISCUSSION THREAD

**Discussion by DetlefR** [arendst/Tasmota#15305] [Ideas]
Hello,
sometimes I get stupid, unbelivable  valus from my TFMini sensor. 
I have inserted a test in **TfmpProcessData** whether the both first bytes are 0x59
`// distance to object (default in cm)
    if (Tfmp_buffer[0]==0x59 && Tfmp_buffer[1]==0x59){
..
..
TfmpSerial->flush();
    }
`
Now the values are trustworthy.
It is possible to insert this check in xsns_86_tfminiplus.ino.
Thanks Detlef

**Comment by arendst** [comment]
Thx for the idea.

Just looked at the driver and was suprised to see that the values are used BEFORE the crc is checked ?!?!

Could you try the below code replacing the current ``TfmpProcessData`` function:

```
/*
void TfmpProcessData(void)
{
    uint16_t crc = 0;
    // distance to object (default in cm)
    tfminiplus_sensor.distance = (Tfmp_buffer[3] << 8) + Tfmp_buffer[2];
    // signal strength (between 100 and 65535 fine, otherwise 0 due to out of range of non-reflective surface)
    tfminiplus_sensor.sigstrength = (Tfmp_buffer[5] << 8) + Tfmp_buffer[4];
    // chip temperature
    tfminiplus_sensor.chiptemp = (((Tfmp_buffer[7] << 8) + Tfmp_buffer[6]) >> 3) - 256;
    // check crc sum
    for (int i = 0; i < TFMP_MAX_DATA_LEN - 1; ++i)
    {
        crc += (uint16_t)Tfmp_buffer[i];
    }
    if (!(char)(crc & 0xff) != Tfmp_buffer[TFMP_MAX_DATA_LEN])
    {
        DEBUG_SENSOR_LOG(PSTR("TFmini Plus: crc error"));
    }
    DEBUG_SENSOR_LOG(PSTR("TFmini Plus: Distance: %d"), tfminiplus_sensor.distance);
    DEBUG_SENSOR_LOG(PSTR("TFmini Plus: Signal: %d"), tfminiplus_sensor.sigstrength);
    DEBUG_SENSOR_LOG(PSTR("TFmini Plus: Chip Temp: %d"), tfminiplus_sensor.chiptemp);
    TfmpSerial->flush();
}
*/
void TfmpProcessData(void) {
    // check crc sum
    uint16_t crc = 0;
    for (int i = 0; i < TFMP_MAX_DATA_LEN - 1; ++i) {
        crc += (uint16_t)Tfmp_buffer[i];
    }
    if ((char)(crc & 0xff) == Tfmp_buffer[TFMP_MAX_DATA_LEN]) {
        // distance to object (default in cm)
        tfminiplus_sensor.distance = (Tfmp_buffer[3] << 8) + Tfmp_buffer[2];
        // signal strength (between 100 and 65535 fine, otherwise 0 due to out of range of non-reflective surface)
        tfminiplus_sensor.sigstrength = (Tfmp_buffer[5] << 8) + Tfmp_buffer[4];
        // chip temperature
        tfminiplus_sensor.chiptemp = (((Tfmp_buffer[7] << 8) + Tfmp_buffer[6]) >> 3) - 256;

        DEBUG_SENSOR_LOG(PSTR("TFmini Plus: Distance: %d"), tfminiplus_sensor.distance);
        DEBUG_SENSOR_LOG(PSTR("TFmini Plus: Signal: %d"), tfminiplus_sensor.sigstrength);
        DEBUG_SENSOR_LOG(PSTR("TFmini Plus: Chip Temp: %d"), tfminiplus_sensor.chiptemp);
    } else {
        DEBUG_SENSOR_LOG(PSTR("TFmini Plus: crc error"));
    }
    TfmpSerial->flush();
}
```

**Reply by DetlefR** [reply]
Hello,

Unfortunately, both of our solutions were not enough.
I found that the main reason for incorrect values are errors on the serial connection.
currentIndex in TfmpAddData counts only from 0 to TFMP_MAX_DATA_LEN and do not respect errors. Sometime currentIndex is out of sync.
I have impemented a check in TfmpAddData. I think the manufacturer know why he use 5959 and this pair will never occur in the data part.
Two other point.  
**if (currentIndex > TFMP_MAX_DATA_LEN)** insted of  **if (currentIndex >= TFMP_MAX_DATA_LEN)** in **TfmpAddData**.
**if ((char)(crc & 0xff) == Tfmp_buffer[TFMP_MAX_DATA_LEN-1]) {** insted of **if ((char)(crc & 0xff) == Tfmp_buffer[TFMP_MAX_DATA_LEN]) {** in **TfmpProcessData**. See // Terminate buffer and reset position in **TfmpAddData**
`

bool TfmpAddData(char nextChar)
{
    // Buffer position
    static uint8_t currentIndex = 0;
    // Store data into buffer at position
    if (currentIndex >0 && (Tfmp_buffer[currentIndex-1] == 0x59 && nextChar == 0x59))
    {
        currentIndex = 1;
    }  
    Tfmp_buffer[currentIndex] = nextChar;
    currentIndex++;
    // Check for too many data
    if (currentIndex > TFMP_MAX_DATA_LEN)
    {
        // Terminate buffer and reset position
        Tfmp_buffer[TFMP_MAX_DATA_LEN] = '\0';
        currentIndex = 0;
        return true;
    }
    return false;
}

`

**Reply by arendst** [reply]
Thx. Just made a development commit with your suggested changes. Pls test if possible (available for download in 25 minutes)

**Reply by DetlefR** [reply]
Hello,

I have tested the new version. 
It is working and robust against error on serial connection.

Thanks

**Comment by OlafWalther1969** [comment]
Hello,
i´ve tried for two weeks to make the TFMINI PLUS Sensor work with TASMOTA. But I haven´t had any success.
Are the changes in the text before now in the developper version for downloading?

Best regards
Olaf

**Reply by Jason2866** [reply]
@OlafWalther1969 yes the changes are in development version.

**Reply by OlafWalther1969** [reply]
Thank you Jason,
I´have downloaded the development version 11.1.0.2 and had compiled it with gitpod.
I´ve flashed the firmware on a wemos d1 mini and a nodemcu v3.
I´ve tried different GPIO for tfmini+ TX and tfmini- RX but nothing has worked.
The sensor works with USB TTL converter.
![1f4d5fbb](https://user-images.githubusercontent.com/66950616/167080516-14dbd26f-4c1c-401c-b706-4129f26abd57.jpg)
![6abf6f42](https://user-images.githubusercontent.com/66950616/167080520-5bc4d9bd-f3ed-4f2d-b165-9d7668b70dbf.jpg)
![43f893d0](https://user-images.githubusercontent.com/66950616/167080523-f483447b-a54e-4912-bbe9-5109b0b3e6de.jpg)

**Comment by Jason2866** [comment]
Since i do not have the device i can not help any further. @arendst can you have a look?

**Reply by OlafWalther1969** [reply]
Thank Jason for your help so far

**Comment by DetlefR** [comment]
Hello,

do you have respected these settings.
From xsns_86_tfminiplus.ino.

==============================================
* Before connecting module to Tasmota, please prepare module with
 * the following sequence through terminal program (e.g. Termite):
 * 1. Connect sensor to USB-to-UART (TTL 3.3V)
 * 2. Set terminal to 115200bps and 8N1
 * 3. Set frame rate to 0 (zero):
 *      send -->: 5A0603000000
 *      receive <-- (OK): 5A0603000000
 * 4. Set baud rate to 9600bps
 *      send -->: 5A0806802500000D
 *      receive <-- (OK): 5A0806802500000D
 * 5. Set terminal to 9600bps and 8N1
 * 6. Save settings
 *      send -->: 5A04116F
 *      receive <-- (OK): 5A05110070
 *      receive <-- (FAIL): 5A05110071
 * 7. When point 6 returned OK, than ready to connect to Tasmota! :)
 *    When point 6 returned FAIL, start with item 1 and
 *      - check connection of RX and TX pin are not exchanged
 *      - check power supply
 *          - 5V +-0,5V
 *          - 110mA average
 *          - 140mA max peak

**Reply by OlafWalther1969** [reply]
Hello Detlef,
Thank you for your quick help!
I´ve installed Termite, set the TTL > USB to 3,3V
Settings Terminate, see picture:

![termite settings](https://user-images.githubusercontent.com/66950616/167111612-cbd52ca4-0259-41bc-8aaa-620deb840a67.jpg)

I´ve not found this setting: "Set frame rate to 0 (zero):"

In Terminal, you can see crypt data continuously, which are responsive to the distance measurement

At this point I struggle

**Comment by DetlefR** [comment]
Hello Olaf,

at first. Set on your terminal under "transmitted text" "Append nothing".
Second. Do you have the possibility on Termite to send and receive Hex code? Maybe on Plug-ins Hex View?
Send 5A0603000000 and hopefully receive 5A0603000000. That means TFMini do not send automatically the distance. Only on request. There should be no more traffic .
 Send 5A0806802500000D. That set the baud rate to 9600. After that you must set the baud rate on Termite to 9600 too.
Send 5A04116F to save these settings.
Try 5A040462. This sends a request for a measurement. 
Important, at least for me. The red cable needs 5V+.

**Reply by OlafWalther1969** [reply]
Hello Detlef,

... first ... Thank you very much for your support and the support from the hole forum!

i´ve made the changes.
sending 5A0603000000 has no effect.
The TFMINI Plus continuously send the measurement data.
I have also tried the benewake terminal software.
I tried set TTL > USB to 3,3V and TTL > USB to 5V
![tfmini_terminal_hex](https://user-images.githubusercontent.com/66950616/167139142-b48ddd42-a4bd-4ad6-bb3b-fc8effe62c4d.jpg)

**Comment by DetlefR** [comment]
Hello Olaf,

it seems you send no hex commands.
Please read this article. [https://softwarerecs.stackexchange.com/questions/35319/serial-console-to-send-hex-bytes-com-port](url)

**Reply by OlafWalther1969** [reply]
Your great Detlef!

Now I´ve got it to work, with your help. I also get well support from benewake.

now I used the original benewake software tool "BW-TFDS V12.2"
for me it was the easiest way to send the hex code to the TFMINI Plus

![photo_2022-05-10_14-01-12](https://user-images.githubusercontent.com/66950616/167632138-da2d7342-05c4-41b6-b776-5c13111f3f41.jpg)

The Hex Code:
Set frame rate to 0 (zero) :  5A 06 03 00 00 00
Set baud rate to 9600bps:  5A 08 06 80 25 00 00 0D
Save settings: 5A 04 11 6F

After sending the hex code you have to repower the sensor

I used the last tasmota development firmware and compiled it with gitpod
Before compiling i comment out "//" for the TFMINI devices

I flashed the tasmota firmware with tasmotizer to the esp8266 (nodemcu amica v3) and set GPIO4 to RD (D2) to "TF MINI+ RX" and GPIO05 (D1) to "TF Mini.

![GPIO](https://user-images.githubusercontent.com/66950616/167635599-f39bce98-bcec-4087-82c8-716d7b5386fe.jpg)

![esp8266](https://user-images.githubusercontent.com/66950616/167634367-e5dc6b73-ae57-4ab5-ac3d-fa8f83bae615.jpg)

![finish](https://user-images.githubusercontent.com/66950616/167634633-88d30e2b-2bac-49d8-bfb5-d6beb9127a19.jpg)

**Reply by RaphDaMan** [reply]
Congrats OlafWalter1969!

**Comment by OlafWalther1969** [comment]
Is the wiring OK?
or could it be a problem using a 5V sensor TX/RX, connecting directly to the gpio of the esp8266? (3,3V?)



![grafik](https://user-images.githubusercontent.com/66950616/167643097-29b9eedd-e5b1-47b3-be1f-9d681cb341e0.png)

**Reply by DetlefR** [reply]
The wiring is OK.
From the datasheet of tfmini. The TX/RX is on 3.3V level.
Are you satisfied with the stability and accuracy of the sensor?

**Reply by RaphDaMan** [reply]
THX for the idea and update!

My next step could be evaluating automatic connection settings when new sensor connected...

I did not expect someone uses this sensor with the awesome Tasmota :)

**Reply by OlafWalther1969** [reply]
@DetlefR The sensor works stable with gpio12/D6/TFMINI+ RX and gpio/D5/TFMINI+ TX.
The accuracy is about 1cm.

I´m using a Wemos Mini D1 ESP8266 12-F.
Now I´m struggle to connect in the same project a LCD 2004 module and a TFMIN+.
I have compiled the development version 11.1.0.2 with gitpod. (included TFMINI and LCD library)

Only on D5/D6/D7 (gpio12/13/14) I´ve got 1 device to work.
Is their a solution to connect one device LCD 2004 (I2C SDA / I2C SCL) and the secound device TFMINI Plus (RX / TX) at one Wemos Mini D1 ESP8266 12-F. I´ve tried many differnt gpios with I2C and RX/TX but I have not found a solution.

best regards

**Reply by RaphDaMan** [reply]
> I´m using a Wemos Mini D1 ESP8266 12-F. Now I´m struggle to connect in the same project a LCD 2004 module and a TFMIN+. I have compiled the development version 11.1.0.2 with gitpod. (included TFMINI and LCD library)
> 
> Only on D5/D6/D7 (gpio12/13/14) I´ve got 1 device to work. Is their a solution to connect one device LCD 2004 (I2C SDA / I2C SCL) and the secound device TFMINI Plus (RX / TX) at one Wemos Mini D1 ESP8266 12-F. I´ve tried many differnt gpios with I2C and RX/TX but I have not found a solution.

I do have an Wemos Mini D1 clone with TFMini+ @GPIO12/13 and AM2301 @GPIO14. It should work fine if you use SDA/SCL @GPIO4/5.
![image](https://user-images.githubusercontent.com/53603025/168885397-5930e50b-e2c6-4080-bccb-f3dfa6e06b1b.png)
Did you check the LCD2004 module address is set properly?

**Comment by DetlefR** [comment]
Hello,

for interesting users of TF Lidar module with I2C interface.
In the meantime, I have written a berry script on ESP32 for this sensor. Find attached.
The script reads every second the distance. It checks if data is ready and the environment conditions (not overexposed) are met.
For the MQTT SENSOR topic, the average is build. On the WebUI the last value is shown.

Configure I2C on ESP32. 
Wire the sensor for I2C mode. The script use the default sensor address 0x10.
The wire 6 is a data ready signal. In the script, the GPIO 23 is used for this.
Rename the script to "tfluna.be". (Github doesn't support .be attachments)
Upload the script from WebUi Consoles->Manage File System->Upload
Create or extend the autoexec.be with this line
load("tfluna.be")

Restart the ESP32

[tfluna.txt](https://github.com/arendst/Tasmota/files/9437771/tfluna.txt)

**Comment by DetlefR** [comment]
Hello,

I test it on Monday.
________________________________
Von: Theo Arends ***@***.***>
Gesendet: Donnerstag, 7. April 2022 14:00
An: arendst/Tasmota ***@***.***>
Cc: DetlefR ***@***.***>; Author ***@***.***>
Betreff: Re: [arendst/Tasmota] xsns_86_tfminiplus.ino (Discussion #15305)


Thx. Just made a development commit with your suggested changes. Pls test if possible (available for download in 25 minutes)
